jena-arq/src/main/java/org/apache/jena/riot/lang/extra/LangParserBase.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.jena.vocabulary.RDF;
+
+/** 
+ * Base for JavaCC parser of RDF languages.
+ * See {@link ParserBase} for the base class for SPARQL. 
+ */
+public class LangParserBase {
+
+    protected final Node XSD_TRUE       = NodeConst.nodeTrue ;
+    protected final Node XSD_FALSE      = NodeConst.nodeFalse ;
+
+    protected final Node nRDFtype       = NodeConst.nodeRDFType ;
+
+    protected final Node nRDFnil        = NodeConst.nodeNil ;
+    protected final Node nRDFfirst      = NodeConst.nodeFirst ;
+    protected final Node nRDFrest       = NodeConst.nodeRest ;
+
+    protected final Node nRDFsubject    = RDF.Nodes.subject ;
+    protected final Node nRDFpredicate  = RDF.Nodes.predicate ;
+    protected final Node nRDFobject     = RDF.Nodes.object ;
+
+    protected StreamRDF stream;
+    protected ParserProfile profile;
+
+    public LangParserBase() { }
+    
+    // These are essential calls unless the parser takes over the functions. 
+    // They can't easily be in the constructor because this class is inherited
+    // by the JaavCC generated parser.
+    public void setProfile(ParserProfile profile) {
+        this.profile = profile;
+    }
+
+    public void setDest(StreamRDF stream) {
+        this.stream = stream;
+    }
+    
+    // ----
+    
+    protected String fixupPrefix(String prefix, int line, int column) {
+        if ( prefix.endsWith(":") )
+            prefix = prefix.substring(0, prefix.length() - 1) ;
+        return prefix ;
+    }
+
+    protected Node createNode(String iriStr) {
+        return profile.createURI(iriStr, -1, -1);
+    }
+
+    protected Node createBNode(int line, int column) {
+        return profile.createBlankNode(null, line, column);
+    }
+
+    protected Node createBNode(String label, int line, int column) {
+        return profile.createBlankNode(null, label, line, column);
+    }
+
+    protected Node createListNode(int line, int column) {
+        return  createBNode(line, column);
+    }
+
+    protected void checkString(String string, int line, int column) {
+        for ( int i = 0 ; i < string.length() ; i++ ) {
+            // Not "codePointAt" which does surrogate processing.
+            char ch = string.charAt(i);
+            // Check surrogate pairs are pairs.
+            if ( Character.isHighSurrogate(ch) ) {
+                i++;
+                if ( i == string.length() )
+                    throw new QueryParseException("Bad surrogate pair (end of string)", line, column);
+                char ch1 = string.charAt(i);
+                if ( ! Character.isLowSurrogate(ch1) ) {
+                    throw new QueryParseException("Bad surrogate pair (high surrogate not followed by low surrogate)", line, column);
+                }
+            } else if ( Character.isLowSurrogate(ch) ) {
+                throw new QueryParseException("Bad surrogate pair (low surrogate without high surrogate)", line, column);
+            }
+        }
+    }
```

Comment:
Was this code elsewhere and it was then moved here? Or is it duplicated? I remember seeing some similar code a few days ago I think.

Reply:
Duplicated; there is some other methods copied as well.

The tiny difference is that the exceptions are RiotParseException ... except I missed this method!  Changed.